File: Preparing for competition/NOIP practice/T45179.cpp
#include<iostream>
#include<string>
using namespace std;


bool search(string word, int N, int M, string matrix[])
{
	//extern matrix;
	int W = word.length();
	bool eq, ok = false;
	// ˙≈≈≤È—Ø
	for(int i = 0; i < N - W + 1; i++)
		for(int j = 0; j < M; j++)
		{
			eq = true;
			for(int k = 0; k < W; k++)
				if(matrix[i + k][j] != word[k]) eq = false;
			if (eq) ok = true;
		} 
	//∫·≈≈≤È—Ø 
	for(int i = 0; i < N; i++)
		for(int j = 0; j < M - W + 1; j++)
		{
			eq = true;
			for(int k = 0; k < W; k++)
				if(matrix[i][j + k] != word[k]) eq = false;
			if (eq) ok = true;
		} 
	return ok;
}
int main()
{
	int N, M;//æÿ’Û≥ﬂ¥Á
	cin >> N >> M;
	string matrix[N]; //æÿ’Û
	for(int i = 0; i < N; i++) 
		cin >> matrix[i];// ‰»Î“ª––
	
	int WNUM;
	cin >> WNUM;
	string In;
	string ans[WNUM]; 
	for(int i = 0; i < WNUM; i++)
	{
		cin >> In;
		ans[i] = (search(In, N, M, matrix) ? "YES" : "NO");
	}
	for(int i = 0; i < WNUM; i++)
		cout << ans[i] << endl;
	return 0;
 } 

File: Preparing for competition/NOIP practice/T45179.py
def search(word, matrix, N, M):
    W = len(word)#单词长度
    #应该判断一下能够横着或者竖着
    #print(word, W)
    ok = False
    '''for i in range(0, N - W + 1):#最右下方的单词开始点：N - W, M - W
        for j in range(0, M - W + 1):
            #纵列比较
            eq = True
            for k in range(W):
                if matrix[i + k][j] != word[k]:
                    eq = False
            if eq: ok = True
            eq = True
            for k in range(W):
                if matrix[i][j + k] != word[k]:
                    eq = False
            if eq: ok = True'''
    for i in range(0, N - W + 1):#竖排查询
        for j in range(0, M):
            eq = True
            for k in range(W):
                if matrix[i + k][j] != word[k]:
                    eq = False
            if eq: ok = True
    for i in range(0, N):#横排查询
        for j in range(0, M - W + 1):
            eq = True
            for k in range(W):
                if matrix[i][j + k] != word[k]:
                    eq = False
            if eq: ok = True
    return ok
N, M = map(int, input().split())#阵列大小

matrix = []
for i in range(N):
    matrix.append(input())
    
for i in matrix: print(i)

WNUM = int(input())
for i in range(WNUM):
    print('YES' if search(input(), matrix, N, M) else 'NO')
